inc/Engine/CriticalPath/RESTGenerate.php

```diff
@@ -0,0 +1,302 @@
+<?php
+
+namespace WP_Rocket\Engine\CriticalPath;
+
+use WP_REST_Request;
+use WP_Rocket\Event_Management\Subscriber_Interface;
+
+class RESTGenerate implements Subscriber_Interface {
+	const ROUTE_NAMESPACE = 'wp-rocket/v1';
+	const API_URL         = 'https://cpcss.wp-rocket.me/api/job/';
+
+	/**
+	 * Base critical CSS path for posts
+	 *
+	 * @var string
+	 */
+	private $critical_css_path;
+
+	/**
+	 * Constructor
+	 *
+	 * @param string $critical_css_path Base critical CSS path.
+	 */
+	public function __construct( $critical_css_path ) {
+		$this->critical_css_path = $critical_css_path . get_current_blog_id() . '/posts/';
+	}
+
+	/**
+	 * Return an array of events that this subscriber wants to listen to.
+	 *
+	 * @since  3.6
+	 *
+	 * @return array
+	 */
+	public static function get_subscribed_events() {
+		return [
+			'rest_api_init' => [
+				[ 'register_generate_route' ],
+			],
+		];
+	}
+
+	/**
+	 * Registers the generate route in the WP REST API
+	 *
+	 * @since 3.6
+	 *
+	 * @return void
+	 */
+	public function register_generate_route() {
+		register_rest_route(
+			self::ROUTE_NAMESPACE,
+			'cpcss/post/(?P<id>[\d]+)',
+			[
+				'methods'             => 'POST',
+				'callback'            => [ $this, 'generate' ],
+				'permission_callback' => [ $this, 'check_permissions' ],
+			]
+		);
+	}
+
+	/**
+	 * Generates the CPCSS for the requested post ID
+	 *
+	 * @since 3.6
+	 *
+	 * @param WP_REST_Request $request WP REST request response.
+	 * @return WP_REST_Response
+	 */
+	public function generate( WP_REST_Request $request ) {
+		$status = get_post_status( $request['id'] );
+
+		if ( ! $status ) {
+			return rest_ensure_response(
+				[
+					'success' => false,
+					'code'    => 'post_not_exists',
+					'message' => __( 'Requested post does not exist.', 'rocket' ),
+					'data'    => [
+						'status' => 400,
+					],
+				]
+			);
+		}
+
+		if ( 'publish' !== $status ) {
+			return rest_ensure_response(
+				[
+					'success' => false,
+					'code'    => 'post_not_published',
+					'message' => __( 'Cannot generate CPCSS for unpublished post.', 'rocket' ),
+					'data'    => [
+						'status' => 400,
+					],
+				]
+			);
+		}
+
+		$post_url  = get_permalink( $request['id'] );
+		$post_type = get_post_type( $request['id'] );
+
+		$response = $this->send_generation_request( $post_url, $post_type );
+
+		if ( false === $response['success'] ) {
+			return rest_ensure_response( $response );
+		}
+
+		while ( $job_data = $this->get_critical_path( $response['data']['id'] ) ) { // phpcs:ignore WordPress.CodeAnalysis.AssignmentInCondition.FoundInWhileCondition
```

Comment:
Why use a `while` here? There is probably a good reason but I can’t see it.

Reply:
It's to loop and re-request to the API as long as the job is not complete. But there might be a better way to do that

Reply:
>re-request to the API as long as the job is not complete

What might drive the job to not be complete that would require another API request for the same data?

Reply:
The way the API is designed, there is a queue system for the job requests. So when we send a GET request to the API for the specific job, it might not have started yet (still in queue behind other jobs), or can also be currently running (`pending` and `generating` statuses).

This is all available under the same endpoint, that's why we need to re-send the request until the endpoint returns a `complete` status in its JSON response (or there is a failure).

Reply:
Hmm, okay that tells me that our communication with the API is not robust or performant. It's forcing us to resend requests over and over and over again.

This gives us the opportunity to sit down and map out a better approach, one where we can queue requests and then get back and ack of its status. Diagramming helps us to visualize and discuss. @Tabrisrp would you like to lead that effort?